projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,182 @@
+	if retryAfterHeaderString == "a while" {
+		return 0, fmt.Errorf("[3] retry-after header is 'a while'")
+	}
+
+	// second condition an integer in string format
+	retryAfterInteger, err := strconv.Atoi(retryAfterHeaderString)
+
+	// if we successfully parse an integer
+	if err == nil {
+		// convert the integer into seconds and return it
+		retryDuration := time.Duration(retryAfterInteger) * time.Second
+		return retryDuration, nil
+	}
+
+	// third condition a timestamp in string format
+	// (!) http.ParseTime can parse various formats
+	// specified in the RFCs for HTTP (RFC1123, RFC850, ANSI C etc)
+	// so it is a better choice than time.Parse(RFC1123, _)
+	retryAfterTimeStamp, err := http.ParseTime(retryAfterHeaderString)
+
+	// if we cannot parse the timestamp then return an error
+	if err != nil {
+		return 0, fmt.Errorf("[3] retry-after header is invalid: %w", err)
+	}
+
+	// if we can parse the timestamp then calculate the difference and return it
+	retryDuration := time.Until(retryAfterTimeStamp)
+
+	return retryDuration, nil
+}
+
+func handleTooManyRequestsResponse(response *http.Response) (time.Duration, error) {
+	// parse the Retry-After Header string
+	retryDuration, err := parseRetryAfterHeader(response.Header.Get("Retry-After"))
+
+	// if we cannot parse the Retry-After Header string then return an error
+	if err != nil {
+		return 0, fmt.Errorf("[2] failed to get a valid delay from retry-after header")
+	}
+
+	// if we reach here we have a retry duration (of type time.Duration)
+	switch {
+	case retryDuration > 5*time.Second:
+		return 0, fmt.Errorf("[2] retry-after is more than 5 seconds")
+	case retryDuration > 0:
+		return retryDuration, nil
+	default:
+		return 0, fmt.Errorf("[2] retry-after condition not found")
+	}
+}
+
+// this is returning a responseBody, a retry duration in seconds, and an error
+// this breaks single responsibility... need to rethink it... suggestions?
+func handleStatusCode(response *http.Response) (string, time.Duration, error) {
+	// handle the various response status codes (extensible for more status codes)
+	switch response.StatusCode {
+	// 200
+	case http.StatusOK:
+		// parse the response body
+		responseBody, err := parseResponseBody(response)
+
+		// if we cannot parse the response body then return an error
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] parseResponseBody failed: %w", err)
+		}
+
+		// otherwise return the parsed response body
+		return responseBody, 0, nil
+
+	// 429
+	case http.StatusTooManyRequests:
+		// get the retry duration
+		retryDuration, err := handleTooManyRequestsResponse(response)
+
+		// if we cannot get a valid retry duration then return an error
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] handleTooManyRequestsResponse failed: %w", err)
+		}
+
+		// return a retry duration
+		return "", retryDuration, nil
+
+	// Other Status Codes
+	default:
+		return "", 0, fmt.Errorf("[1] unhandled response status code: %d", response.StatusCode)
```

Comment:
I'd consider including the response body here in case it says something useful, but I think it's also completely fine not to.